### ISSUE TITLE
Export a combination of flatAlt and group

### ISSUE THREAD

**Issue by sjakobi** [quchen/prettyprinter#138]
While looking at `dhall`'s [usage](https://github.com/dhall-lang/dhall-haskell/blob/6ccf08d6fb28ee35cf9e938cdce225e65ea84df6/dhall/src/Dhall/Pretty/Internal.hs) of `prettyprinter`, I noticed that `flatAlt` is always combined with `group`.

(The only direct use of `group` where the argument is not a `flatAlt`-expression is on the top-level [here](https://github.com/dhall-lang/dhall-haskell/blob/6ccf08d6fb28ee35cf9e938cdce225e65ea84df6/dhall/src/Dhall/Pretty/Internal.hs#L513).)

What I'm currently thinking about is something like

```
prefer :: Doc ann -> Doc ann -> Doc ann
prefer a b = group (flatAlt b a)
```

The order of arguments is reversed in comparison to `flatAlt` since in `prefer a b`, we "prefer" `a` over `b`.

Thoughts? Any ideas for a better name?!

**Comment by quchen** [comment]
I can see the use case, but I’m very careful introducing new combinations of existing functions into the library. There are too many functions in there already, making it hard to keep everything offered in mind. What’s the difference between `hcat`, `sep`, `cat`?

At some level, it’s unfortunate that the library exposes functions of different API levels – `group` is fairly low-level, while `vcat` for example is more semantic.

So in conclusion, I’m usually against adding super simple combinations of existing functions.

**Comment by sjakobi** [comment]
Many thanks for the feedback @quchen! :)

The distinction between high-level on low-level combinators is very interesting! I think I'll try to categorize the current API along that axis for myself.

If `group` is low-level, I think `prefer` could also serve as a good high-level demonstration of `group` (and also `flatAlt`).

I agree with the principle of not carelessly adding new things to `Prettyprint.Doc` though. Although I do wish I had a good guideline for what's worth including and what isn't!

**Comment by sjakobi** [comment]
> I agree with the principle of not carelessly adding new things to `Prettyprint.Doc` though. Although I do wish I had a good guideline for what's worth including and what isn't!

I have since learned about https://wiki.haskell.org/Fairbairn_threshold.
